lib/benchee/benchmark/runner.ex

```diff
@@ -238,131 +179,39 @@ defmodule Benchee.Benchmark.Runner do
-  defp main_function(function, input), do: fn -> function.(input) end
-
-  # Builds the appropriate function to benchmark. Takes into account the
-  # combinations of the following cases:
-  #
-  # * an input is specified - creates a 0-argument function calling the original
-  #   function with that input
-  # * number of iterations - when there's more than one iteration we repeat the
-  #   benchmarking function during execution and measure the the total run time.
-  #   We only run multiple iterations if a function is so fast that we can't
-  #   accurately measure it in one go. Hence, we can't split up the function
-  #   execution and hooks anymore and sadly we also measure the time of the
-  #   hooks.
-  defp build_benchmarking_function(
-         %Scenario{
-           function: function,
-           before_each: nil,
-           after_each: nil
-         },
-         %ScenarioContext{
-           num_iterations: iterations,
-           scenario_input: input,
-           config: %{after_each: nil, before_each: nil}
-         }
-       )
-       when iterations > 1 do
-    main = main_function(function, input)
-    # with no before/after each we can safely omit them and don't get the hit
-    # on run time measurements (See PR discussions for this for more info #127)
-    fn -> RepeatN.repeat_n(main, iterations) end
-  end
-
-  defp build_benchmarking_function(
-         scenario = %Scenario{function: function},
-         scenario_context = %ScenarioContext{num_iterations: iterations}
-       )
-       when iterations > 1 do
-    fn ->
-      RepeatN.repeat_n(
-        fn ->
-          new_input = run_before_each(scenario, scenario_context)
-          main = main_function(function, new_input)
-          return_value = main.()
-          run_after_each(return_value, scenario, scenario_context)
-        end,
-        iterations
+  def measure(
+        scenario,
+        scenario_context = %ScenarioContext{
+          num_iterations: iterations
+        },
+        measurer
       )
-    end
+      when iterations > 1 do
+    RepeatedMeasurement.measure(scenario, scenario_context, measurer)
   end
 
-  defp run_before_each(
-         %{
-           before_each: local_before_each
-         },
-         %{
-           config: %{before_each: global_before_each},
-           scenario_input: input
-         }
-       ) do
-    input
-    |> run_before_function(global_before_each)
-    |> run_before_function(local_before_each)
-  end
-
-  defp run_after_each(
-         return_value,
-         %{
-           after_each: local_after_each
-         },
-         %{
-           config: %{after_each: global_after_each}
-         }
-       ) do
-    if local_after_each, do: local_after_each.(return_value)
-    if global_after_each, do: global_after_each.(return_value)
-  end
+  @no_input Benchmark.no_input()
+  def main_function(function, @no_input), do: function
+  def main_function(function, input), do: fn -> function.(input) end
```

Comment:
I had hoped to get rid off this together and saving it in the `ScenarioContext` - something I wanted to do for a long time as I also fear that this might introduce a performance overhead/stands in the way of optimizations.

However, as hooks can change the input it's not as easy. Now that we have our own measurers we could just call them with an argument if we feel like it and implement them like this. I'd wanna leave this for another day though :)